File: typescript/tests/aether_contracts.test.ts
// aether_contracts.test.ts
//
// Verifies the CircleAI.Aether port (5 contracts) + deterministic in-memory impls:
//   event records + derived predicates (IsExit/HopCount/ExceedsLoss/IsHighSeverity/…)
//   enum ordinals (wire contract)
//   InMemoryAetherTelemetry (snapshot-safe fan-out, dispose)
//   InMemoryAetherContext (isAvailable / isSufficient / requiresAuth)
//   AuthChallengeResult factories + InMemoryAuthChallenge (minimum enforcement)
//   InMemoryAetherIntelligence (health / assessment / routing / trust stream)
//   SecurityDirective predicates (HasTarget / IsPermanent)

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  // enums
  AetherNodeEventKind,
  AetherTransportKind,
  AetherTransportEventKind,
  AetherRouteEventKind,
  AetherSecurityEventKind,
  AetherThreatLevel,
  AetherNetworkEventKind,
  AetherInstallLevel,
  SecurityDirectiveKind,
  AuthChallengeReason,
  AuthMethod,
  // predicates
  aetherNodeEventIsExit,
  aetherNodeHealthIsValid,
  aetherTransportEventExceedsLoss,
  aetherRouteEventHopCount,
  aetherRouteEventIsFailed,
  aetherSecurityEventIsHighSeverity,
  aetherNetworkEventIsHighCongestion,
  networkHealthReportIsValid,
  threatAssessmentIsValid,
  trustScoreUpdateHasChanged,
  trustScoreUpdateIsDegraded,
  securityDirectiveHasTarget,
  securityDirectiveIsPermanent,
  // versions
  aetherVersion,
  compareAetherVersion,
  // impls
  NullAetherTelemetry,
  InMemoryAetherTelemetry,
  InMemoryAetherContext,
  AuthChallengeResult,
  InMemoryAuthChallenge,
  InMemoryAetherIntelligence,
  // types
  type IAetherTelemetryObserver,
  type AetherNodeEvent,
  type AetherTransportEvent,
  type AetherRouteEvent,
  type AetherSecurityEvent,
  type AetherNetworkEvent,
  type SecurityDirective,
} from "../src/aether/index";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function nodeEvt(kind: AetherNodeEventKind, trust = 1.0): AetherNodeEvent {
  return {
    nodeId: "n1",
    kind,
    health: { trustScore: trust, isReachable: true, latencyMs: 10, hopCount: 1 },
    occurredAt: new Date(),
  };
}

function secEvt(kind: AetherSecurityEventKind, level: AetherThreatLevel, node = "n1"): AetherSecurityEvent {
  return { nodeId: node, kind, threatLevel: level, description: `${kind}@${node}`, metadata: {}, occurredAt: new Date() };
}

class CollectObserver implements IAetherTelemetryObserver {
  nodes: AetherNodeEvent[] = [];
  security: AetherSecurityEvent[] = [];
  transports: AetherTransportEvent[] = [];
  routes: AetherRouteEvent[] = [];
  networks: AetherNetworkEvent[] = [];
  onNodeEvent(e: AetherNodeEvent): void {
    this.nodes.push(e);
  }
  onSecurityEvent(e: AetherSecurityEvent): void {
    this.security.push(e);
  }
  onTransportEvent(e: AetherTransportEvent): void {
    this.transports.push(e);
  }
  onRouteEvent(e: AetherRouteEvent): void {
    this.routes.push(e);
  }
  onNetworkEvent(e: AetherNetworkEvent): void {
    this.networks.push(e);
  }
}

// ---------------------------------------------------------------------------
// Enum ordinals — wire contract
// ---------------------------------------------------------------------------

describe("Aether enum ordinals", () => {
  it("match the C# declaration order", () => {
    assert.equal(AetherNodeEventKind.Joined, 0);
    assert.equal(AetherNodeEventKind.HealthChanged, 2);
    assert.equal(AetherTransportKind.WiFi, 0);
    assert.equal(AetherTransportKind.Unknown, 6);
    assert.equal(AetherTransportEventKind.PacketLoss, 3);
    assert.equal(AetherRouteEventKind.Failed, 2);
    assert.equal(AetherSecurityEventKind.NodeAuthAttempt, 0);
    assert.equal(AetherSecurityEventKind.PrivilegeAttempt, 5);
    assert.equal(AetherThreatLevel.None, 0);
    assert.equal(AetherThreatLevel.Critical, 4);
    assert.equal(AetherNetworkEventKind.PartitionDetected, 2);
    assert.equal(AetherInstallLevel.None, 0);
    assert.equal(AetherInstallLevel.OS, 2);
    // SecurityDirectiveKind order: UpdateNodeTrust, AvoidNode, QuarantineNode, ReleaseNode, RequestReauth, ElevateMonitoring
    assert.equal(SecurityDirectiveKind.UpdateNodeTrust, 0);
    assert.equal(SecurityDirectiveKind.ReleaseNode, 3);
    assert.equal(SecurityDirectiveKind.ElevateMonitoring, 5);
    assert.equal(AuthChallengeReason.OsLevelToggle, 0);
    assert.equal(AuthChallengeReason.ManualRequest, 4);
    // AuthMethod is strength-ordered starting at 1.
    assert.equal(AuthMethod.Biometric, 1);
    assert.equal(AuthMethod.BiometricAndDeviceAdmin, 3);
    assert.equal(AuthMethod.Custom, 4);
  });
});

// ---------------------------------------------------------------------------
// Event predicates
// ---------------------------------------------------------------------------

describe("Aether event predicates", () => {
  it("node IsExit / health IsValid", () => {
    assert.equal(aetherNodeEventIsExit(nodeEvt(AetherNodeEventKind.Left)), true);
    assert.equal(aetherNodeEventIsExit(nodeEvt(AetherNodeEventKind.Joined)), false);
    assert.equal(aetherNodeHealthIsValid({ trustScore: 0.5, isReachable: true, latencyMs: 0, hopCount: 0 }), true);
    assert.equal(aetherNodeHealthIsValid({ trustScore: 1.5, isReachable: true, latencyMs: 0, hopCount: 0 }), false);
  });

  it("transport ExceedsLoss", () => {
    const e: AetherTransportEvent = {
      nodeId: "n",
      kind: AetherTransportEventKind.PacketLoss,
      transport: AetherTransportKind.WiFi,
      latencyMs: null,
      packetLossRate: 0.4,
      occurredAt: new Date(),
    };
    assert.equal(aetherTransportEventExceedsLoss(e, 0.3), true);
    assert.equal(aetherTransportEventExceedsLoss(e, 0.5), false);
    assert.equal(aetherTransportEventExceedsLoss({ ...e, packetLossRate: null }, 0.1), false);
  });

  it("route HopCount / IsFailed", () => {
    const e: AetherRouteEvent = {
      sourceNodeId: "a",
      destinationNodeId: "c",
      path: ["a", "b", "c"],
      kind: AetherRouteEventKind.Failed,
      failureReason: "loop",
      occurredAt: new Date(),
    };
    assert.equal(aetherRouteEventHopCount(e), 3);
    assert.equal(aetherRouteEventIsFailed(e), true);
    assert.equal(aetherRouteEventIsFailed({ ...e, kind: AetherRouteEventKind.Discovered }), false);
  });

  it("security IsHighSeverity", () => {
    assert.equal(aetherSecurityEventIsHighSeverity(secEvt(AetherSecurityEventKind.IntrusionSignal, AetherThreatLevel.High)), true);
    assert.equal(aetherSecurityEventIsHighSeverity(secEvt(AetherSecurityEventKind.IntrusionSignal, AetherThreatLevel.Critical)), true);
    assert.equal(aetherSecurityEventIsHighSeverity(secEvt(AetherSecurityEventKind.NodeAuthAttempt, AetherThreatLevel.Low)), false);
  });

  it("network IsHighCongestion", () => {
    const base: AetherNetworkEvent = {
      kind: AetherNetworkEventKind.CongestionDetected,
      nodeCount: 5,
      activeRouteCount: 3,
      congestionLevel: 0.8,
      occurredAt: new Date(),
    };
    assert.equal(aetherNetworkEventIsHighCongestion(base), true);
    assert.equal(aetherNetworkEventIsHighCongestion({ ...base, congestionLevel: 0.75 }), false);
  });
});

// ---------------------------------------------------------------------------
// Intelligence-output record predicates
// ---------------------------------------------------------------------------

describe("Intelligence record predicates", () => {
  it("report/assessment IsValid", () => {
    assert.equal(
      networkHealthReportIsValid({ overallScore: 0.9, trustedNodeCount: 1, suspiciousNodeCount: 0, summary: "", generatedAt: new Date() }),
      true,
    );
    assert.equal(
      networkHealthReportIsValid({ overallScore: 1.1, trustedNodeCount: 0, suspiciousNodeCount: 0, summary: "", generatedAt: new Date() }),
      false,
    );
    assert.equal(
      threatAssessmentIsValid({ nodeId: "n", threatConfidence: 0.2, level: AetherThreatLevel.Low, indicators: [], assessedAt: new Date() }),
      true,
    );
  });

  it("TrustScoreUpdate HasChanged / IsDegraded", () => {
    const u = { nodeId: "n", previousScore: 0.9, currentScore: 0.6, reason: "x", updatedAt: new Date() };
    assert.equal(trustScoreUpdateHasChanged(u), true);
    assert.equal(trustScoreUpdateIsDegraded(u), true);
    const same = { ...u, currentScore: 0.9 };
    assert.equal(trustScoreUpdateHasChanged(same), false);
    assert.equal(trustScoreUpdateIsDegraded(same), false);
    // Below the 0.001 movement floor.
    assert.equal(trustScoreUpdateHasChanged({ ...u, currentScore: 0.9005 }), false);
  });
});

// ---------------------------------------------------------------------------
// SecurityDirective predicates
// ---------------------------------------------------------------------------

describe("SecurityDirective predicates", () => {
  const dir = (target: string | null, durationMs: number | null): SecurityDirective => ({
    kind: SecurityDirectiveKind.AvoidNode,
    targetNodeId: target,
    trustScoreOverride: null,
    threatLevel: AetherThreatLevel.High,
    reason: "r",
    durationMs,
    issuedAt: new Date(),
  });

  it("HasTarget / IsPermanent", () => {
    assert.equal(securityDirectiveHasTarget(dir("n1", null)), true);
    assert.equal(securityDirectiveHasTarget(dir(null, null)), false);
    assert.equal(securityDirectiveHasTarget(dir("   ", null)), false);
    assert.equal(securityDirectiveIsPermanent(dir("n1", null)), true);
    assert.equal(securityDirectiveIsPermanent(dir("n1", 1000)), false);
  });
});

// ---------------------------------------------------------------------------
// Telemetry
// ---------------------------------------------------------------------------

describe("NullAetherTelemetry", () => {
  it("returns a no-op disposable and never throws", () => {
    const obs = new CollectObserver();
    const handle = NullAetherTelemetry.instance.subscribe(obs);
    handle.dispose();
    handle.dispose(); // idempotent
    assert.equal(obs.nodes.length, 0);
  });
});

describe("InMemoryAetherTelemetry", () => {
  it("fans out to all subscribers", () => {
    const hub = new InMemoryAetherTelemetry();
    const a = new CollectObserver();
    const b = new CollectObserver();
    hub.subscribe(a);
    hub.subscribe(b);
    assert.equal(hub.subscriberCount, 2);

    hub.publishNode(nodeEvt(AetherNodeEventKind.Joined));
    hub.publishSecurity(secEvt(AetherSecurityEventKind.IntrusionSignal, AetherThreatLevel.Critical));
    assert.equal(a.nodes.length, 1);
    assert.equal(b.nodes.length, 1);
    assert.equal(a.security.length, 1);
    assert.equal(b.security.length, 1);
  });

  it("dispose unsubscribes exactly one observer", () => {
    const hub = new InMemoryAetherTelemetry();
    const a = new CollectObserver();
    const b = new CollectObserver();
    const ha = hub.subscribe(a);
    hub.subscribe(b);
    ha.dispose();
    assert.equal(hub.subscriberCount, 1);
    hub.publishNode(nodeEvt(AetherNodeEventKind.Joined));
    assert.equal(a.nodes.length, 0);
    assert.equal(b.nodes.length, 1);
  });

  it("survives a subscriber disposing inside its own callback (snapshot-safe)", () => {
    const hub = new InMemoryAetherTelemetry();
    let handleRef: { dispose(): void } | null = null;
    const selfRemover: IAetherTelemetryObserver = {
      onNodeEvent() {
        handleRef?.dispose();
      },
      onSecurityEvent() {},
      onTransportEvent() {},
      onRouteEvent() {},
      onNetworkEvent() {},
    };
    const tail = new CollectObserver();
    handleRef = hub.subscribe(selfRemover);
    hub.subscribe(tail);
    // Must not throw; tail must still receive the event.
    hub.publishNode(nodeEvt(AetherNodeEventKind.Joined));
    assert.equal(tail.nodes.length, 1);
    assert.equal(hub.subscriberCount, 1);
  });
});

// ---------------------------------------------------------------------------
// Version + Context
// ---------------------------------------------------------------------------

describe("AetherVersion", () => {
  it("compares component-by-component", () => {
    assert.ok(compareAetherVersion(aetherVersion(2, 6), aetherVersion(2, 5)) > 0);
    assert.ok(compareAetherVersion(aetherVersion(2, 5), aetherVersion(2, 5)) === 0);
    assert.ok(compareAetherVersion(aetherVersion(1, 9), aetherVersion(2, 0)) < 0);
  });
});

describe("InMemoryAetherContext", () => {
  it("App level, enabled → available, no auth needed", () => {
    const ctx = new InMemoryAetherContext({ installLevel: AetherInstallLevel.App, runtimeVersion: aetherVersion(2, 6) });
    assert.equal(ctx.isAvailable, true);
    assert.equal(ctx.isEnabled, true);
    assert.equal(ctx.requiresAuth, false);
    assert.equal(ctx.isSufficient, true); // null minimum → always sufficient
  });

  it("None level → not available", () => {
    const ctx = new InMemoryAetherContext({ installLevel: AetherInstallLevel.None });
    assert.equal(ctx.isAvailable, false);
  });

  it("disabled OS instance → not available but requiresAuth", () => {
    const ctx = new InMemoryAetherContext({ installLevel: AetherInstallLevel.OS, isEnabled: false });
    assert.equal(ctx.requiresAuth, true);
    assert.equal(ctx.isAvailable, false);
  });

  it("isSufficient honours the minimum version", () => {
    const ok = new InMemoryAetherContext({ runtimeVersion: aetherVersion(2, 6), minimumRequired: aetherVersion(2, 5) });
    assert.equal(ok.isSufficient, true);
    const low = new InMemoryAetherContext({ runtimeVersion: aetherVersion(2, 4), minimumRequired: aetherVersion(2, 5) });
    assert.equal(low.isSufficient, false);
    const absent = new InMemoryAetherContext({ runtimeVersion: null, minimumRequired: aetherVersion(2, 5) });
    assert.equal(absent.isSufficient, false);
  });
});

// ---------------------------------------------------------------------------
// Auth challenge
// ---------------------------------------------------------------------------

describe("AuthChallengeResult", () => {
  it("success / failure factories", () => {
    const s = AuthChallengeResult.success(AuthMethod.BiometricAndDeviceAdmin);
    assert.equal(s.succeeded, true);
    assert.equal(s.failureReason, null);
    assert.equal(s.methodUsed, AuthMethod.BiometricAndDeviceAdmin);
    const f = AuthChallengeResult.failure(AuthMethod.Biometric, "nope");
    assert.equal(f.succeeded, false);
    assert.equal(f.failureReason, "nope");
  });
});

describe("InMemoryAuthChallenge", () => {
  it("defaults the minimum to BiometricAndDeviceAdmin when null", async () => {
    const auth = new InMemoryAuthChallenge(AuthMethod.BiometricAndDeviceAdmin);
    const r = await auth.challengeAsync(AuthChallengeReason.PrivilegedOperation, null, "confirm");
    assert.equal(r.succeeded, true);
    assert.equal(r.methodUsed, AuthMethod.BiometricAndDeviceAdmin);
    assert.equal(auth.issued[0].minimum, AuthMethod.BiometricAndDeviceAdmin);
  });

  it("fails when the available method is weaker than the requested minimum", async () => {
    const auth = new InMemoryAuthChallenge(AuthMethod.Biometric); // only biometric
    const r = await auth.challengeAsync(AuthChallengeReason.PrivilegedOperation, AuthMethod.BiometricAndDeviceAdmin, "x");
    assert.equal(r.succeeded, false);
    assert.match(r.failureReason ?? "", /weaker than required/);
  });

  it("OS toggle always enforces BiometricAndDeviceAdmin at minimum", async () => {
    const weak = new InMemoryAuthChallenge(AuthMethod.Biometric);
    const denied = await weak.requestOsToggleAsync(true);
    assert.equal(denied.succeeded, false);
    assert.equal(weak.issued[0].reason, AuthChallengeReason.OsLevelToggle);
    assert.equal(weak.issued[0].minimum, AuthMethod.BiometricAndDeviceAdmin);

    const strong = new InMemoryAuthChallenge(AuthMethod.BiometricAndDeviceAdmin);
    const ok = await strong.requestOsToggleAsync(false);
    assert.equal(ok.succeeded, true);
  });

  it("honours a scripted denyReason", async () => {
    const auth = new InMemoryAuthChallenge(AuthMethod.Custom, "user cancelled");
    const r = await auth.challengeAsync(AuthChallengeReason.ManualRequest, AuthMethod.Biometric, "x");
    assert.equal(r.succeeded, false);
    assert.equal(r.failureReason, "user cancelled");
  });
});

// ---------------------------------------------------------------------------
// InMemoryAetherIntelligence
// ---------------------------------------------------------------------------

describe("InMemoryAetherIntelligence", () => {
  it("empty mesh reports full health", async () => {
    const intel = new InMemoryAetherIntelligence();
    const h = await intel.getNetworkHealthAsync();
    assert.equal(h.overallScore, 1.0);
    assert.equal(h.trustedNodeCount, 0);
    assert.equal(h.summary, "No nodes observed.");
  });

  it("security events degrade trust and drive assessment + routing", async () => {
    const intel = new InMemoryAetherIntelligence();
    // Two Critical intrusion signals on n-bad: 0.15 * 3 = 0.45 each → 1.0 → 0.55 → 0.10
    intel.recordSecurityEvent(secEvt(AetherSecurityEventKind.IntrusionSignal, AetherThreatLevel.Critical, "n-bad"));
    intel.recordSecurityEvent(secEvt(AetherSecurityEventKind.IntrusionSignal, AetherThreatLevel.Critical, "n-bad"));

    const a = await intel.assessThreatAsync("n-bad");
    assert.equal(a.level, AetherThreatLevel.Critical); // score ~0.10 ≤ 0.25
    assert.ok(a.threatConfidence > 0.8);
    assert.ok(a.indicators.includes("intrusion-signal-detected"));
    assert.ok(a.indicators.includes("high-severity-event"));

    const advice = await intel.getRoutingAdviceAsync("n-bad");
    assert.deepEqual(advice.recommendedPath, []); // below avoid threshold
    assert.ok(advice.avoidNodes.includes("n-bad"));

    // Unknown node → full trust, direct path.
    const good = await intel.getRoutingAdviceAsync("n-good");
    assert.deepEqual(good.recommendedPath, ["n-good"]);
  });

  it("streams trust updates buffered before the reader attaches (unbounded)", async () => {
    const intel = new InMemoryAetherIntelligence();
    // Publish BEFORE subscribing — unbounded channel must retain these.
    intel.setTrust("n1", 0.4, "manual-a");
    intel.setTrust("n1", 0.2, "manual-b");

    const received: string[] = [];
    const controller = new AbortController();
    const consumer = (async () => {
      for await (const u of intel.streamTrustScoresAsync(controller.signal)) {
        received.push(u.reason);
        if (received.length === 2) {
          controller.abort();
          break;
        }
      }
    })();
    await consumer;
    assert.deepEqual(received, ["manual-a", "manual-b"]);
  });
});

File: rust/tests/aether_contracts_test.rs
//! aether_contracts_test.rs
//!
//! Ports the behavioural contracts in `CircleAI.Aether/*.cs`: the event records'
//! computed properties, `AetherVersion` (.NET `System.Version`) ordering,
//! `StaticAetherContext` / `AetherNetContextAdapter` derivations, the auth-
//! challenge minimum-method policy, the security-layer directive publishing +
//! posture, and the intelligence outputs.

use std::collections::BTreeMap;
use std::sync::{Arc, Mutex};

use chrono::{Duration, Utc};
use circle_ai::aether::{
    AetherInstallLevel, AetherNetworkEvent, AetherNetworkEventKind, AetherNodeEvent,
    AetherNodeEventKind, AetherNodeHealth, AetherRouteEvent, AetherRouteEventKind,
    AetherSecurityEvent, AetherSecurityEventKind, AetherThreatLevel, AetherTransportEvent,
    AetherTransportEventKind, AetherTransportKind, AetherVersion, AuthChallengeReason, AuthMethod,
    IAISecurityLayer, IAetherContext, IAetherIntelligence, IAetherTelemetry, IAuthChallenge,
    ISecurityDirectiveConsumer, InMemoryAISecurityLayer, InMemoryAetherIntelligence,
    InMemoryAetherTelemetry, NullAetherTelemetry, PolicyAuthChallenge, SecurityDirective,
    SecurityDirectiveKind, StaticAetherContext,
};

// ── Event record computed properties ─────────────────────────────────────────

#[test]
fn node_event_is_exit_only_on_left() {
    let health = AetherNodeHealth::new(1.0, true, Duration::milliseconds(10), 1);
    let joined = AetherNodeEvent::new("n1", AetherNodeEventKind::Joined, health.clone(), Utc::now());
    let left = AetherNodeEvent::new("n1", AetherNodeEventKind::Left, health.clone(), Utc::now());
    assert!(!joined.is_exit());
    assert!(left.is_exit());
    assert!(health.is_valid());
    assert!(!AetherNodeHealth::new(1.5, true, Duration::zero(), 0).is_valid());
}

#[test]
fn transport_event_exceeds_loss() {
    let e = AetherTransportEvent::new(
        "n1",
        AetherTransportEventKind::PacketLoss,
        AetherTransportKind::WiFi,
        Some(Duration::milliseconds(5)),
        Some(0.30),
        Utc::now(),
    );
    assert!(e.exceeds_loss(0.25));
    assert!(!e.exceeds_loss(0.50));
    // None packet-loss never exceeds.
    let none = AetherTransportEvent::new(
        "n1",
        AetherTransportEventKind::Selected,
        AetherTransportKind::LoRa,
        None,
        None,
        Utc::now(),
    );
    assert!(!none.exceeds_loss(0.0));
}

#[test]
fn route_event_hop_count_and_failed() {
    let e = AetherRouteEvent::new(
        "a",
        "c",
        vec!["a".into(), "b".into(), "c".into()],
        AetherRouteEventKind::Failed,
        Some("timeout".into()),
        Utc::now(),
    );
    assert_eq!(e.hop_count(), 3);
    assert!(e.is_failed());
}

#[test]
fn security_event_high_severity() {
    let mk = |lvl| {
        AetherSecurityEvent::new(
            "n1",
            AetherSecurityEventKind::IntrusionSignal,
            lvl,
            "d",
            BTreeMap::new(),
            Utc::now(),
        )
    };
    assert!(!mk(AetherThreatLevel::Medium).is_high_severity());
    assert!(mk(AetherThreatLevel::High).is_high_severity());
    assert!(mk(AetherThreatLevel::Critical).is_high_severity());
}

#[test]
fn network_event_high_congestion() {
    let hi = AetherNetworkEvent::new(AetherNetworkEventKind::CongestionDetected, 5, 3, 0.80, Utc::now());
    let lo = AetherNetworkEvent::new(AetherNetworkEventKind::TopologyChanged, 5, 3, 0.50, Utc::now());
    assert!(hi.is_high_congestion());
    assert!(!lo.is_high_congestion());
}

// ── AetherVersion ordering (System.Version semantics) ────────────────────────

#[test]
fn version_ordering_matches_dotnet() {
    // Two-component build=-1 sorts before three-component build=0.
    assert!(AetherVersion::new(1, 0) < AetherVersion::with_build(1, 0, 0));
    assert!(AetherVersion::new(2, 1) > AetherVersion::new(2, 0));
    assert!(AetherVersion::with_build(3, 0, 5) > AetherVersion::with_build(3, 0, 4));
    assert_eq!(AetherVersion::new(1, 2), AetherVersion::new(1, 2));
    assert!(AetherVersion::full(1, 0, 0, 1) > AetherVersion::with_build(1, 0, 0));
    assert_eq!(AetherVersion::with_build(1, 2, 3).to_string(), "1.2.3");
    assert_eq!(AetherVersion::new(1, 2).to_string(), "1.2");
}

// ── IAetherContext ───────────────────────────────────────────────────────────

#[test]
fn static_context_derivations() {
    // OS-managed, sufficient version, enabled.
    let ctx = StaticAetherContext::new(
        AetherInstallLevel::Os,
        Some(AetherVersion::with_build(3, 0, 0)),
        Some(AetherVersion::with_build(2, 5, 0)),
        true,
    );
    assert!(ctx.is_available());
    assert!(ctx.is_sufficient());
    assert!(ctx.requires_auth()); // OS level
    assert!(ctx.is_enabled());

    // Absent context.
    let absent = StaticAetherContext::absent();
    assert!(!absent.is_available());
    assert!(absent.is_sufficient()); // no minimum required → always sufficient
    assert!(!absent.requires_auth());
}

#[test]
fn context_insufficient_when_runtime_below_minimum() {
    let ctx = StaticAetherContext::new(
        AetherInstallLevel::App,
        Some(AetherVersion::with_build(1, 0, 0)),
        Some(AetherVersion::with_build(2, 0, 0)),
        true,
    );
    assert!(!ctx.is_sufficient());
    assert!(!ctx.requires_auth()); // App level, not OS
}

// ── IAuthChallenge ────────────────────────────────────────────────────────────

#[test]
fn auth_challenge_enforces_os_floor() {
    // A user who can only do Biometric fails the OS floor (Biometric+DeviceAdmin).
    let weak = PolicyAuthChallenge::biometric_only();
    let r = weak.request_os_toggle(true);
    assert!(!r.succeeded);
    assert!(r.failure_reason.is_some());

    // A user who can do the full method passes.
    let strong = PolicyAuthChallenge::always_succeeds();
    let r2 = strong.request_os_toggle(false);
    assert!(r2.succeeded);
    assert_eq!(r2.method_used, AuthMethod::Custom);
}

#[test]
fn auth_challenge_requested_minimum_can_only_raise_bar() {
    // Exactly meeting the floor with BiometricAndDeviceAdmin succeeds for a
    // non-OS reason with no explicit minimum.
    let user = PolicyAuthChallenge::new(AuthMethod::BiometricAndDeviceAdmin);
    let ok = user.challenge(AuthChallengeReason::PrivilegedOperation, None, "prompt");
    assert!(ok.succeeded);

    // Requesting a stronger Custom minimum now fails that same user.
    let fail = user.challenge(
        AuthChallengeReason::PrivilegedOperation,
        Some(AuthMethod::Custom),
        "prompt",
    );
    assert!(!fail.succeeded);
}

#[test]
fn auth_method_strength_ordering() {
    assert!(AuthMethod::Custom > AuthMethod::BiometricAndDeviceAdmin);
    assert!(AuthMethod::BiometricAndDeviceAdmin > AuthMethod::DeviceAdmin);
    assert!(AuthMethod::DeviceAdmin > AuthMethod::Biometric);
    assert_eq!(AuthMethod::Biometric as u8, 1);
    assert_eq!(AuthMethod::Custom as u8, 4);
}

// ── IAISecurityLayer (InMemoryAISecurityLayer) ───────────────────────────────

/// Records every directive it receives.
#[derive(Default)]
struct RecordingConsumer {
    got: Mutex<Vec<SecurityDirective>>,
}
impl ISecurityDirectiveConsumer for RecordingConsumer {
    fn on_directive(&self, directive: &SecurityDirective) {
        self.got.lock().unwrap().push(directive.clone());
    }
}

fn sec_event(node: &str, lvl: AetherThreatLevel) -> AetherSecurityEvent {
    AetherSecurityEvent::new(
        node,
        AetherSecurityEventKind::IntrusionSignal,
        lvl,
        "sig",
        BTreeMap::new(),
        Utc::now(),
    )
}

#[test]
fn security_layer_publishes_directive_on_high_and_critical() {
    let telemetry = InMemoryAetherTelemetry::new();
    let layer = InMemoryAISecurityLayer::new();
    let consumer = Arc::new(RecordingConsumer::default());
    let _sub = layer.subscribe_to_directives(consumer.clone());

    layer.start(&telemetry);
    assert!(layer.is_active());

    // Medium → ElevateMonitoring.
    telemetry.publish_security_event(&sec_event("n1", AetherThreatLevel::Medium));
    // High → AvoidNode.
    telemetry.publish_security_event(&sec_event("n1", AetherThreatLevel::High));
    // Critical → QuarantineNode.
    telemetry.publish_security_event(&sec_event("n1", AetherThreatLevel::Critical));

    let got = consumer.got.lock().unwrap();
    let kinds: Vec<SecurityDirectiveKind> = got.iter().map(|d| d.kind).collect();
    assert_eq!(
        kinds,
        vec![
            SecurityDirectiveKind::ElevateMonitoring,
            SecurityDirectiveKind::AvoidNode,
            SecurityDirectiveKind::QuarantineNode,
        ]
    );
    for d in got.iter() {
        assert_eq!(d.target_node_id.as_deref(), Some("n1"));
        assert!(d.has_target());
    }
}

#[test]
fn security_layer_subscribe_before_publish_is_not_lost() {
    // The observer must be wired synchronously by start(); an event published
    // immediately after start must be seen.
    let telemetry = InMemoryAetherTelemetry::new();
    let layer = InMemoryAISecurityLayer::new();
    layer.start(&telemetry);
    telemetry.publish_security_event(&sec_event("n1", AetherThreatLevel::High));
    let posture = layer.get_posture();
    assert_eq!(posture.overall_threat_level, AetherThreatLevel::High);
    assert_eq!(posture.monitored_node_count, 1);
    assert!(posture.is_active);
}

#[test]
fn security_layer_posture_counts_and_node_exit_clears() {
    let telemetry = InMemoryAetherTelemetry::new();
    let layer = InMemoryAISecurityLayer::new();
    layer.start(&telemetry);

    telemetry.publish_security_event(&sec_event("crit", AetherThreatLevel::Critical));
    telemetry.publish_security_event(&sec_event("med", AetherThreatLevel::Medium));

    let p = layer.get_posture();
    assert_eq!(p.quarantined_node_count, 1);
    assert_eq!(p.monitored_node_count, 1);
    assert_eq!(p.overall_threat_level, AetherThreatLevel::Critical);

    // A node departure clears its tracked threat.
    let health = AetherNodeHealth::new(0.0, false, Duration::zero(), 0);
    telemetry.publish_node_event(&AetherNodeEvent::new(
        "crit",
        AetherNodeEventKind::Left,
        health,
        Utc::now(),
    ));
    let p2 = layer.get_posture();
    assert_eq!(p2.quarantined_node_count, 0);
    assert_eq!(p2.overall_threat_level, AetherThreatLevel::Medium);

    layer.stop();
    assert!(!layer.is_active());
    assert!(!layer.get_posture().is_active);
}

#[test]
fn security_layer_stop_unsubscribes_observer() {
    let telemetry = InMemoryAetherTelemetry::new();
    let layer = InMemoryAISecurityLayer::new();
    layer.start(&telemetry);
    assert_eq!(telemetry.subscriber_count(), 1);
    layer.stop();
    assert_eq!(telemetry.subscriber_count(), 0, "stop drops the subscription");
}

#[test]
fn null_telemetry_subscribe_is_noop() {
    let t = NullAetherTelemetry::new();
    let layer = InMemoryAISecurityLayer::new();
    layer.start(&t); // must not panic
    // Nothing is published; posture is clean.
    assert_eq!(layer.get_posture().overall_threat_level, AetherThreatLevel::None);
}

// ── IAetherIntelligence (InMemoryAetherIntelligence) ─────────────────────────

#[test]
fn intelligence_empty_network_health() {
    let intel = InMemoryAetherIntelligence::new();
    let h = intel.get_network_health();
    assert_eq!(h.overall_score, 1.0);
    assert_eq!(h.trusted_node_count, 0);
    assert!(h.is_valid());
}

#[test]
fn intelligence_assess_unknown_is_zero_confidence() {
    let intel = InMemoryAetherIntelligence::new();
    let a = intel.assess_threat("ghost");
    assert_eq!(a.threat_confidence, 0.0);
    assert_eq!(a.level, AetherThreatLevel::None);
    assert!(a.indicators.is_empty());
}

#[test]
fn intelligence_low_trust_drives_level_and_avoid() {
    let intel = InMemoryAetherIntelligence::new();
    intel.set_trust("good", 0.95, "seed");
    intel.set_trust("bad", 0.20, "attack");

    let assess = intel.assess_threat("bad");
    assert_eq!(assess.level, AetherThreatLevel::Critical);
    assert!(assess.threat_confidence > 0.75);
    assert!(assess.indicators.contains(&"trust-critical".to_string()));

    let advice = intel.get_routing_advice("bad");
    assert!(advice.recommended_path.is_empty(), "no safe path to quarantined node");
    assert!(advice.avoid_nodes.contains(&"bad".to_string()));

    let advice2 = intel.get_routing_advice("good");
    assert_eq!(advice2.recommended_path, vec!["good".to_string()]);
}

#[test]
fn intelligence_stream_drains_updates() {
    let intel = InMemoryAetherIntelligence::new();
    intel.set_trust("n1", 0.9, "a");
    intel.set_trust("n1", 0.5, "b"); // changed → 2 updates
    let first = intel.stream_trust_scores();
    assert_eq!(first.len(), 2);
    assert!(first[1].is_degraded());
    assert!(first[0].has_changed());
    // Second drain is empty.
    assert!(intel.stream_trust_scores().is_empty());
    // A no-op set (same value) publishes nothing.
    intel.set_trust("n1", 0.5, "noop");
    assert!(intel.stream_trust_scores().is_empty());
}
